### PR TITLE
Peer/Score strategies: ranges all the way down

### PR DESCRIPTION
**Pull request by rf**
~~Currently breaks pool-of-servers because it expects perfect distribution.~~

### CODE REVIEW

File: peer.js [hunk 1]
```diff
@@ -626,42 +629,36 @@ TChannelPeer.prototype.makeOutConnection = function makeOutConnection(socket) {
...
+// This had the attribute that a less loaded peer was merely more likely to
+// be chosen over a more loaded peer.
+// We observed with the introduction of a heap, that a less favored peer
+// would have its score less frequently re-evaluated.
+// An emergent behavior was that scores would, over time, be squeezed
+// toward zero and the least favored peer would remain the least favored
+// for ever increasing durations.
+//
+// This remains true with this algorithm, within each equivalence class.
+TChannelPeer.prototype.pendingWeightedRange = function pendingWeightedRange() {
```
Comment (jcorbin) [inline peer.js]: Now With More Inlinability ™; good call.

File: peer.js [hunk 2]
```diff
@@ -692,9 +689,39 @@ function _maybeInvalidateScore(reason) {
     }
 };
 
+TChannelPeer.prototype.getScoreRange = function getScoreRange() {
+    var self = this;
+
+    if (!self._range) {
```
Comment (jcorbin) [inline peer.js]: Let's not add caching in this pr please, defer for later work.

Also, if we do go with this mode of caching, it should be contained in the getScoreRange method entirely:
- computeScoreRange should just compute and return
- the set of self._range should be in getScoreRange

File: sub_peers.js [hunk 3]
```diff
@@ -129,7 +129,7 @@ TChannelSubPeers.prototype.chooseLinearPeer = function chooseLinearPeer(req) {
         var hostPort = hosts[i];
         var peer = self._map[hostPort];
         if (!req || !req.triedRemoteAddrs || !req.triedRemoteAddrs[hostPort]) {
-            var score = peer.scoreStrategy.getScore(req);
+            var score = peer.getScore(req);
```
Comment (jcorbin) [inline sub_peers.js]: hmm, it looks to me like we could separate out the change to the relation ship of peer and strategy from the switch to range api... thouhts?

File: range.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+'use strict';
+
+module.exports = Range;
+
+function Range(lo, hi) {
+    var self = this;
+
+    self.lo = lo;
+    self.hi = hi;
+}
```
Comment (jcorbin) [inline range.js]: please add `#inspect` and `#toString` methods, ala #123

**Comment by jcorbin** [comment]
Why does it break pool-of-servers balance? It shouldn't at all... perhaps that's because of the caching you slipped in to this one without invalidation?

**Comment by rf** [comment]
No longer breaks the distribution, it turned out to just be a bug in the range logic.

**Comment by rf** [comment]
Now ready for review @Raynos @jcorbin

**Comment by jcorbin** [comment]
lgtm, but let's wait for @Raynos to weight in

**Comment by Raynos** [comment]
#sick Seems legit.

**Pull request merged**